File: SEM1_BIO_PROJECT.py
import csv
import time

print("")
print("")
print("===============================GENE FINDER PROGRAM===============================")
print("credits: ritvik, srihari, anubhav, shanmukha, sibiram and aditya")
print("=================================================================================")
time.sleep(1)
print("This program reads your CSV file in the same directory[dnaseq.csv] for dna data")
print("=====================================READING=====================================")
time.sleep(2)
print("")
print("")
def genefinder(dnaseq):
    start_codon = "ATG" 
    stop_codons = ["TAA", "TAG", "TGA"]  
    genes = []  
    start_gene, stop_gene = 0, 3
    
    while start_gene < len(dnaseq):  
        start_gene = dnaseq.find(start_codon, start_gene)
        if start_gene == -1:
            return ["No gene in the given sequence"]
        
        for stop_codon in stop_codons:
            stop_gene = dnaseq.find(stop_codon, start_gene + 3)
            if (stop_gene != -1) and (stop_gene - start_gene) % 3 == 0:
                genes.append(dnaseq[start_gene:stop_gene + 3])
                start_gene = stop_gene + 3
                stop_gene = start_gene + 3
                break
        else:
            start_gene += 1
            stop_gene += 1

    return genes

with open("dnaseq.csv", mode="r") as file:
    reader = csv.DictReader(file)
    for row in reader:
        dnaseq = row["enter_data_line_by_line"]
        genes = genefinder(dnaseq)
        time.sleep(1)
        print(f"The genes identified in the sequence '{dnaseq}' are:")
        for gene in genes:
            print(gene)
        print()  
print("=================================================================================")
time.sleep(1)
print("Edit the CSV file containing the dna sequence to change input parameters")

File: SEM1_BIO_PROJECT.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h> //for that sleep function ig

#define maxlinelen 1024
#define maxgenelen 100

void genefinder(const char *dnaseq, char genes[maxgenelen][maxlinelen], int *geneCount) 
{
    const char *start_codon = "ATG";
    const char *stop_codons[] = {"TAA", "TAG", "TGA"};
    *geneCount = 0;

    int start_gene = 0;
    int len = strlen(dnaseq);

    while (start_gene < len) 
    {
        char *start_pos = strstr(dnaseq + start_gene, start_codon);
        if (start_pos == NULL) 
        {
            if (*geneCount == 0) 
            {
                strcpy(genes[0], "No gene in the given sequence");
                *geneCount = 1;
            }
            return;
        }

        start_gene = start_pos - dnaseq;

        for (int i = 0; i < 3; i++) 
        {
            char *stop_pos = strstr(dnaseq + start_gene + 3, stop_codons[i]);
            if (stop_pos != NULL) 
            {
                int stop_gene = stop_pos - dnaseq; 
                if ((stop_gene - start_gene) % 3 == 0) 
                {
                    strncpy(genes[*geneCount], dnaseq + start_gene, stop_gene - start_gene + 3);
                    genes[*geneCount][stop_gene - start_gene + 3] = '\0'; 
                    (*geneCount)++;
                    start_gene = stop_gene + 3; 
                    break;
                }
            }
        }

        start_gene++;
    }
}

int main() {
    printf("\n\n");
    printf("===============================GENE FINDER PROGRAM===============================\n");
    printf("credits: ritvik, srihari, anubhav, shanmukha, sibiram and aditya\n");
    printf("=================================================================================\n");
    sleep(1);
    printf("This program reads your CSV file in the same directory [dnaseq.csv] for DNA data\n");
    printf("=====================================READING=====================================\n");
    sleep(2);
    printf("\n\n");

    FILE *file = fopen("dnaseq.csv", "r");
    if (file == NULL) 
    {
        perror("Error opening file");
        return 1;
    }

    char line[maxlinelen];
    char genes[maxgenelen][maxlinelen];
    int geneCount;

    // Read the header line
    if (fgets(line, sizeof(line), file) == NULL) 
    {
        perror("Error reading file");
        fclose(file);
        return 1;
    }

    // Process each line
    while (fgets(line, sizeof(line), file)) 
    {
        // Remove trailing newline
        line[strcspn(line, "\n")] = '\0';

        printf("Processing DNA sequence: %s\n", line);
        genefinder(line, genes, &geneCount);

        sleep(1);
        printf("The genes identified in the sequence '%s' are:\n", line);
        for (int i = 0; i < geneCount; i++) {
            printf("%s\n", genes[i]);
        }
        printf("\n");
    }

    fclose(file);

    printf("=================================================================================\n");
    sleep(1);
    printf("Edit the CSV file containing the DNA sequence to change input parameters\n");

    return 0;
}

